### PR TITLE
Preprocessing layers RFC

### PR DESCRIPTION
**Pull request by fchollet**
# Keras Preprocessing Layers

**Comment period is now closed. It was open until May 17th, 2019.**

| Status        | Proposed       |
:-------------- |:---------------------------------------------------- |
| **Author(s)** | Mark Omernick (momernick@google.com), Stan Bileschi (bileschi@google.com), Kester Tong (kestert@google.com), Francois Chollet (fchollet@google.com) |
| **Updated**   | 2019-05-21                                           |
| **Status**   |  Accepted                                           |

## Objective

We aim at providing additional Keras layers to handle [data preprocessing operations](https://en.wikipedia.org/wiki/Data_pre-processing)
such as text vectorization, data normalization, and data discretization (binning).
These operations are currently handled separately from a Keras model via utilities such
as those from `keras.preprocessing`.

These new layers will allow users to include data preprocessing directly in their Keras model, so as to create models that map raw data (such as uint8 tensors for images, or string tensors for text) to predictions.

### CODE REVIEW

File: rfcs/20190502-preprocessing-layers.md [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+
+Case where a user has a single preprocessing layer to do text vectorization where each input sample is encoded as a sequence of word indices.
+
+```python
+vectorization = keras.layers.TextVectorization(mode='int')
+vectorization.adapt(data_sample)
+
+model = keras.Sequential([
+    vectorization,
+    keras.layers.Emebdding(128),  # The number of int indices is not specified since it is inferred.
```
Comment (TimZaman) [inline rfcs/20190502-preprocessing-layers.md]: Emebdding

File: rfcs/20190502-preprocessing-layers.md [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+        self.submodel = MySubmodel()
+
+    def call(self, inputs):
+        return self.submodel(self.preproc_layer)
+
+    def get_preprocessing_stage(self):
+        return self.preproc_layer
+
+    def get_main_stage(self):
+        return self.submodel
```
Comment (Dref360) [inline rfcs/20190502-preprocessing-layers.md:475]: By default this would return `self`?

File: rfcs/20190502-preprocessing-layers.md [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+# model containing the `Dense` layer only.
+main_stage = model.get_main_stage()
+```
+
+Functional model with 2 branches, each with a preprocessing layer:
+
+```python
+normalization_a = layers.Normalization()
+normalization_b = layers.Normalization()
+normalization_a.adapt(data_a)
```
Comment (Dref360) [inline rfcs/20190502-preprocessing-layers.md:436]: I guess this would also support Dataset, Sequence, etc like fit_generator?

File: rfcs/20190502-preprocessing-layers.md [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+### Handling of async prefetching
+
+Some preprocessing ops are CPU-only and benefit from being executed asynchronously on the accelerator host (as opposed to the accelerator itself, e.g. GPU or TPU),
+with a batch of data being prepocessed on the host while the previous batch is being processed by the accelerator. This pattern is known as "async prefetching".
+
+This is normally done as part of a tf.data pipeline. The current proposal implies moving some of that preprocessing to inside the model itself, which is normally
+executed end-to-end on an accelerator.
+
+This means that we need a way to lift the preprocessing part of the model in a tf.data pipeline during model training. In `fit`, we can do this automatically.
+In custom training loops, we will expect the user to do it manually.
```
Comment (Dref360) [inline rfcs/20190502-preprocessing-layers.md]: Could you add an example?

**Review by Dref360 — COMMENTED** [review]
Great work!

File: rfcs/20190502-preprocessing-layers.md (hunk 3 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:436]: Ideally yes, but since this method has to be overridden by layer implementers, there are complications.

File: rfcs/20190502-preprocessing-layers.md (hunk 4 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md]: There is an example below ("Custom training loops" subsection).

File: rfcs/20190502-preprocessing-layers.md (hunk 2 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:475]: By default, it would be NotImplemented on a subclassed model.

File: rfcs/20190502-preprocessing-layers.md [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+Cons: It's very generic, and does not clearly convey the difference between "preprocessing stage" and "main processing stage" required by the async prefetching API.
+
+
+#### Name to use for `adapt` method
+
+We may want to use the name `fit` instead (other suggestions welcome).
+
+Pros of using `fit`: consistency with `model.fit()`, and the `fit` method on `ImageDataGenerator` and `Tokenizer` from the `keras.preprocessing` module.
+Cons of using `fit`: It may confuse users, since `preprocessing_layer.fit()` would have a different signature.
+
```
Comment (gabrieldemarmiesse) [inline rfcs/20190502-preprocessing-layers.md:583]: Can we add a point here about what the Scikit-learn API does? I believe consistency with existing libraries APIs is important and avoids documentation lookups. That would be in the pros.

**Review by gabrieldemarmiesse — COMMENTED** [review]
I love this RFC.

File: rfcs/20190502-preprocessing-layers.md (hunk 5 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:583]: The equivalent Scikit-Learn classes (called `*Vectorizer`) use `fit`, e.g.  [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) -- despite having a different signature compared to `model.fit()`.

This would be an argument in favor of using `fit` here as well.

File: rfcs/20190502-preprocessing-layers.md [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+normed_b = normalization_b(input_b)
+a = layers.Dense(32)(normed_a)
+b = layers.Dense(32)(normed_b)
+c = layers.concatenate([a, b])
+outputs = layers.Dense(1, activation='sigmoid')(c)
+
+model = Model([input_a, input_b], outputs)
+
+# `PreprocessingStage` instance
+# mapping `[input_a, input_b]` to `[normed_a, normed_b]`
```
Comment (gabrieldemarmiesse) [inline rfcs/20190502-preprocessing-layers.md:451]: `Normalization` can benefit from being done in the accelerator. 
* Without the graph optimization, the computing time of the `Normalization` is negligible compared to, let's say, a `Dense` or `Convolutional` layer, because it's an elementwise operation done on a fairly small tensor.
* With a graph optimization, it can be folded into the first layer, since first layers are often linear operations. Making the preprocessing time null. This may not be done with async prefetching. I might be wrong about this second point, I guess it depends how much the graph is optimized behind the scenes.

Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:451]: This is a good point in general: the hypothesis that a preprocessing layer should be moved to async prefetching isn't always true. We'll have to keep this in mind.

In the case of `Normalization`, I would think it is cheap enough that it wouldn't matter whether we run it on the accelerator or on the host.

File: rfcs/20190502-preprocessing-layers.md (hunk 5 above)
Comment (AakashKumarNain) [reply rfcs/20190502-preprocessing-layers.md:583]: The word `fit` is more appropriate IMO as it is consistent with other libraries as well. Though the `argument` that it can confuse some users is fair but I believe most of the users(not including beginners here) would be able to well differentiate between the two. For example, If `fit` is used on `preprocessing`, then you know it is data preprocessing and have extra things which `model.fit()` may or may not have in common

File: rfcs/20190502-preprocessing-layers.md (hunk 3 above)
Comment (AakashKumarNain) [reply rfcs/20190502-preprocessing-layers.md:436]: Does this mean that the whole `data` should be in memory first?

**Comment by isaacgerg** [comment]
Could you add a batch 2D FFT layer?

**Comment by isaacgerg** [comment]
You could also start from the perturbations in "Benchmarking Neural Network Robustness to Common Corruptions and Perturbations" (https://github.com/hendrycks/robustness/)

**Comment by gabrieldemarmiesse** [comment]
Two other questions come to mind. 
* Let's say one wants to use the Normalisation layer on a pre-trained model. This person already knows the mean and the variances of the RGB channels. So this person should not need to use `adapt() `. What is the recommended way of doing it?
* Is there a will to use those preprocessing layers in keras_applications? Just curious.

**Comment by AakashKumarNain** [comment]
> Two other questions come to mind.
> 
> * Let's say one wants to use the Normalisation layer on a pre-trained model. This person already knows the mean and the variances of the RGB channels. So this person should not need to use `adapt() `. What is the recommended way of doing it?
> * Is there a will to use those preprocessing layers in keras_applications? Just curious.

This a good point actually

File: rfcs/20190502-preprocessing-layers.md (hunk 3 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:436]: Not necessarily, it is passed as a dataset, then the `adapt` method will iterate on the dataset in batches.

**Comment by fchollet** [comment]
> Is there a will to use those preprocessing layers in keras_applications? Just curious.

It would be a breaking change, so probably not in existing applications, at least not immediately. In future applications, definitely.

> Let's say one wants to use the Normalisation layer on a pre-trained model. This person already knows the mean and the variances of the RGB channels. So this person should not need to use adapt() . What is the recommended way of doing it?

I would recommend calling `.set_weights([r, g, b])` on the normalization layer of the loaded model, and never calling `adapt`. This is the same as for any other layer.

**Comment by fchollet** [comment]
> Could you add a batch 2D FFT layer?

There are no plans for additional layers at this time. Future layers may be added via TF Addons or via specialized packages such as tf.text.

**Comment by gabrieldemarmiesse** [comment]
>  In future applications, definitely.

Very nice :)

File: rfcs/20190502-preprocessing-layers.md [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+```python
+class MyModel(Model):
+
+    def __init__(self, **kwargs):
+        super(MyModel, self).__init__(**kwargs)
+        self.preproc_layer = layers.Normalization()
+        self.submodel = MySubmodel()
+
+    def call(self, inputs):
+        return self.submodel(self.preproc_layer)
```
Comment (bileschi) [inline rfcs/20190502-preprocessing-layers.md]: typo: we forgot inputs should be 
` return self.submodel(self.preproc_layer(inputs))`

File: rfcs/20190502-preprocessing-layers.md [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+    """
+```
+
+#### TextVectorization
+
+This layer has basic options for managing text in the Keras model.
+It is expected that more advanced users needing custom control will uses Keras-compatible layers provided by tf.text.
+
+Transform a batch of strings (one sample = one string) into either a list of token indices
+(one sample = 1D int tensor or a dense representation (1 sample = 1D float vector).
```
Comment (bileschi) [inline rfcs/20190502-preprocessing-layers.md]: nit missing close paren
`(one sample = 1D int tensor) or a dense representation (1 sample = 1D float vector)`.

File: rfcs/20190502-preprocessing-layers.md [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+             tokens=None,
+             standardize='lower_and_strip_punctuation',
+             split='whitespace',
+             ngrams=1,
+             mode='int',
+             max_length=None):
+    """Transforms text into dense vectors or sequences of word indices.
+
+    Arguments:
+        tokens: None (default) | int | list<string>
```
Comment (edloper) [inline rfcs/20190502-preprocessing-layers.md:262]: You're using "tokens" to mean a list of possible string types.  This isn't how the word "token" has traditionally been used/defined in the NLP field. It was initially introduced to stress the token/type distinction (https://en.wikipedia.org/wiki/Type%E2%80%93token_distinction) when handling text -- e.g., "the dog saw the cat" has 5 word tokens, but 4 word types.  E.g., see https://nlp.stanford.edu/IR-book/html/htmledition/tokenization-1.html or https://www.nltk.org/book/ch01.html#counting-vocabulary.

I therefore think "tokens" should be renamed to "vocabulary" or perhaps "terms."

File: rfcs/20190502-preprocessing-layers.md [hunk 10]
```diff
@@ -0,0 +1,578 @@
...
+
+```python
+preprocessing_layer.adapt(data_sample)
+```
+
+**Possible non-differentiability**
+
+Processing layers extend Keras by allowing preprocessing to be part of the model. Unlike existing layers, these computations are not always differentiable, e.g. both `Discretize` and `VectorizeText` are non-differentiable.
+
+As a result, all preprocessing layers are treated as frozen when used as part of a model. In addition, if a non-differentiable layer is used in the middle of a model (rather than at the start), the model will raise an exception related to differentiability when trying to compute gradients (e.g. as part of `fit`).
```
Comment (sandeep-krishnamurthy) [inline rfcs/20190502-preprocessing-layers.md:113]: I am assuming here that this validation will be at compile time.

File: rfcs/20190502-preprocessing-layers.md [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+
+There are two ways to instantiate a `PrepocessingStage` layer: either `Sequential` style (pass a list of preprocessing layer instances) or Functional style (pass the inputs and outputs of a DAG of preprocessing layers).
+
+If any layer other than `PreprocessingLayer` instances is included in a `PrepocessingStage`, these layers will be treated as frozen both during `adapt` and later during `fit`.
+
+
+#### Normalization
+
+```python
+def __init__(self, axis=-1, **kwargs):
```
Comment (sandeep-krishnamurthy) [inline rfcs/20190502-preprocessing-layers.md:157]: I think users should optionally be able to specify `mean` and `std` parameters to be used for this layer? Default can be adapting to the given dataset, but, it should also be possible for me to say what mean and std to use.

File: rfcs/20190502-preprocessing-layers.md [hunk 12]
```diff
@@ -0,0 +1,578 @@
...
+            preprocessing stages).
+        reset_state: Optional argument specifying whether to clear the state of the
+            layer at the start of the call to `adapt`, or whether to start from
+            the existing state. This argument may not be relevant to all
+            preprocessing layers: a subclass of PreprocessingLayer may chose to
+            only implement `adapt(self, data)`.
+    """
+```
+
+#### PrepocessingStage
```
Comment (sandeep-krishnamurthy) [inline rfcs/20190502-preprocessing-layers.md:147]: Two more most commonly used preprocessing layers are:
1. Resize
2. ToTensor => Converting an image tensor (0 to 255) into a tensor (0 to 1) values.

Should we support it in this proposal?

File: rfcs/20190502-preprocessing-layers.md [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+    This is the part of the model that should be executed asynchronously
+    on the device host during training.
+
+    Returns:
+        Instance of `PreprocessingLayer` or `PreprocessingStage`.
+        May be None if the model does not start with preprocessing layers.
+    """
+    pass
+
+def get_main_stage(self):
```
Comment (sandeep-krishnamurthy) [inline rfcs/20190502-preprocessing-layers.md:387]: nit: IMO main_stage may not be very intuitive, should we call it model_stage?

File: rfcs/20190502-preprocessing-layers.md (hunk 12 above)
Comment (sandeep-krishnamurthy) [reply rfcs/20190502-preprocessing-layers.md:147]: Also, these 3 pre-processing operations (ToTensor, Resize, Normalize) are very well suited to be handed off to the accelerator for significant performance boost. 

I had done a similar work in MXNet Gluon - https://medium.com/apache-mxnet/fusing-data-transformations-with-neural-network-for-faster-inference-and-simpler-production-2be50457be8a

**Review by sandeep-krishnamurthy — COMMENTED** [review]
Thank you for this nice proposal. Few comments.

File: rfcs/20190502-preprocessing-layers.md (hunk 11 above)
Comment (gabrieldemarmiesse) [reply rfcs/20190502-preprocessing-layers.md:157]: I believe this was answered in https://github.com/keras-team/governance/pull/1#issuecomment-490259942 . Are you thinking about another use case?

Comment (sandeep-krishnamurthy) [reply rfcs/20190502-preprocessing-layers.md:157]: Thanks. Yes answers.

File: rfcs/20190502-preprocessing-layers.md (hunk 12 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:147]: Your `ToTensor` seems like it would be covered by the normalization layer (it's just a divide by `255` for each channel).

The issue with a layer that does resizing is that it could not process a batch of tensors. All layers (including preprocessing layers) are meant for batch-wise processing: one batch of N samples in, one batch of N samples out.

The problem here being that if you have images of different sizes, you couldn't pack them into a batch tensor.  So by the time images enter a layer, they should all have the same size. So resizing should happen in the tf.data pipeline.

It could work to resize a batch of images that have already been standardized to a common size, but that is not very useful.

File: rfcs/20190502-preprocessing-layers.md (hunk 10 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:113]: Probably at the start of `fit`/etc. You don't need gradients for `evaluate` or `predict` so raising in `compile` is too early (we don't actually query gradients in `compile` today).

File: rfcs/20190502-preprocessing-layers.md (hunk 12 above)
Comment (sandeep-krishnamurthy) [reply rfcs/20190502-preprocessing-layers.md:147]: 1. ToTensor => Agreed, can be put in Normalization layer. But, it should be an option for users? If not, here we are assuming it will be always a an image with values 0..255.
2. Yes this is a big user problem, during inference specially, that they need to manage resizing. How about, resize preprocessing layer that can take list of tensors and resize and batchify the output into a batched tensor. By solving this, I think we will make the model fully self-sufficient to just take an image and give back results.

File: rfcs/20190502-preprocessing-layers.md (hunk 9 above)
Comment (fchollet) [reply rfcs/20190502-preprocessing-layers.md:262]: This will be discussed in the design review. My take, in summary, is two-fold:

- The concern with using "vocabulary" or "terms" is that it implicitly refers to words, whereas we mean atoms to be arbitrary substrings (including ngrams, subwords, and even characters). Meanwhile it is well accepted that "tokens" can be arbitrary atomic substrings; subwords 
 are a popular choice, and the literature even refers to ["character-level tokenization"](https://www.google.com/search?q=site%3Aarxiv.org+"character-level+tokenization").

- The distinction between atom instances (tokens) and atom types is not confusing for users, and in NLP software the term "token" seems to be often used to refer to unique atoms (types), perhaps because in either case they are encoded as the exact same Python string (so a string instance like "the cat" is both a token, if found in a document, and a type, if used as a key in a dict, but it's the same string). If you pass a list of atomic strings to a tokenizer constructor, it is clear that these refer to unique types. It's not dissimilar to the fact that e.g. `2` refers simultaneously to to the abstract int "2" type, and to specific instances of a 2.

If you're going to call the process "tokenization" and refer to the atoms that you split your string into as "tokens", then it's perfectly reasonable to pass a "tokens" argument in the constructor of the layer.

**Comment by FedericoV** [comment]
How would a binning layer work without doing a first pass over the first?  Any binning layer would want to encode low frequency words/categorical variables into an 'unk' token - but the frequency is very difficult to estimate a priori without doing a first pass over the data.
